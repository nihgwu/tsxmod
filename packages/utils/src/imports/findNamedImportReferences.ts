import { Project, Node, ts } from 'ts-morph'

/**
 * Find all references for a named import.
 *
 * @example
 * const references = findNamedImportReferences(project, 'package', 'Stack')
 */

export function findNamedImportReferences(
  project: Project,
  moduleSpecifierValue: string,
  namedImportName: string
) {
  let references: Node[] = []

  for (const sourceFile of project.getSourceFiles()) {
    for (const importDeclaration of sourceFile.getImportDeclarations()) {
      if (
        importDeclaration.getModuleSpecifierValue() === moduleSpecifierValue
      ) {
        for (const namedImport of importDeclaration.getNamedImports()) {
          if (namedImport.getName() === namedImportName) {
            const identifier = namedImport.getFirstDescendantByKindOrThrow(
              ts.SyntaxKind.Identifier
            )

            references = identifier.findReferencesAsNodes()

            break
          }
        }
      }
    }
  }

  return references
}
