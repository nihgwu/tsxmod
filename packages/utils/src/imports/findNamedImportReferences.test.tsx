import { readFileSync } from 'fs'
import { Project } from 'ts-morph'
import { findNamedImportReferences } from './findNamedImportReferences'

const dialogSource = readFileSync(
  'fixtures/app-with-library/Dialog.tsx',
  'utf8'
)
const appSource = readFileSync('fixtures/app-with-library/App.tsx', 'utf8')

test('get references for named import', () => {
  const project = new Project({
    useInMemoryFileSystem: true,
    compilerOptions: {
      baseUrl: '.',
      paths: {
        'theme-ui': ['theme-ui.d.ts'],
      },
    },
  })

  project.createSourceFile('Dialog.tsx', dialogSource)
  project.createSourceFile('App.tsx', appSource)
  project.createSourceFile(
    'theme-ui.d.ts',
    `
      export const Button: any
      export const Card: any
    `
  )

  const references = findNamedImportReferences(project, 'theme-ui', 'Button')

  expect(
    references.map((reference) => reference.getParentOrThrow().getText())
  ).toMatchSnapshot()
})
