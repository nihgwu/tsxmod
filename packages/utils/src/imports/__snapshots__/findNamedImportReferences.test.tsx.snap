// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get references for named import 1`] = `
[
  "Button",
  "<Button onClick={props.onClose} variant="secondary">",
  "</Button>",
  "Button: any",
  "Button",
  "<Button onClick={() => setOpen(true)} children="Open" />",
]
`;
